rust_src/src/multibyte.rs

```diff
@@ -88,13 +92,85 @@ impl LispStringRef {
+    /// width is measured by how many columns it occupies on the screen while
+    /// paying attention to compositions. This is a convenience function for
+    /// `self.display_width(None)`
     pub fn width(self) -> usize {
-        unsafe { lisp_string_width(self.into(), -1, ptr::null_mut(), ptr::null_mut()) as usize }
+        let (width, _) = self.display_width(None);
+        width
+    }
+
+    /// Return width of the string when displayed in the current buffer. The
+    /// width is measured by how many columns it occupies on the screen while
+    /// paying attention to compositions.
+    ///
+    /// With `precision` argument, return the width of longest substring that
+    /// doesn't exceed `precision`, and the number of characters and bytes it
+    /// contains in the returned tuple.
+    pub fn display_width(self, precision: Option<usize>) -> (usize, Option<(usize, usize)>) {
+        // Manually determine if string is unibyte.
+        let len = self.len_chars() as usize;
+        let multibyte = self.len_chars() < self.len_bytes();
+        let distab = unsafe { buffer_display_table() };
+        // Sum width
+        let mut width = 0;
+        // Character index
+        let mut i = 0;
+        // Byte index
+        let mut b = 0;
+
+        while i < len {
+            let (chars, bytes, thiswidth) = unsafe {
+                // If there is a composition, get its id.
+                let (cmp_id, end) = match find_composition(i, None, self.into()) {
+                    Some((_, end, val)) => (
+                        get_composition_id(
+                            i as isize,
+                            b as isize,
+                            (end - i) as isize,
+                            val,
+                            self.into(),
+                        ),
+                        end,
+                    ),
+                    None => (-1, 0),
+                };
+                if cmp_id >= 0 {
+                    // Character is a composition, look it up in the composition table.
+                    let chars = end - i;
+                    let bytes = string_char_to_byte(self.into(), end as isize) - b as isize;
+                    let thiswidth = (*(*composition_table.offset(cmp_id))).width as usize;
+                    (chars, bytes as usize, thiswidth)
+                } else {
+                    // Character is a single codepoint, calculate it if multibyte, otherwise get
+                    // raw byte at b.
+                    let (c, bytes) = if multibyte {
+                        string_char_and_length(self.const_data_ptr().add(b))
+                    } else {
+                        (self.as_slice()[b].into(), 1)
+                    };
+                    (1, bytes, char_width(c as i32, distab) as usize)
+                }
+            };
+
+            // Return if adding character exceeds precision
+            if let Some(precision) = precision {
```

Comment:
This early return check just relies on `thiswidth`, right? And that's calculated independently of `chars` and `bytes`? If so, you could calculate `thiswidth` first, then run the early return check, then calculate `chars` and `bytes` and save yourself a little extra work if it does return early.

Reply:
I think calculating thiswidth first scatters the core logic of the function too far apart for the slight optimisation to be worth the readability compromise